Adds balancing for reactions involving compounds

LalitChauhan56 opened the pull request:
Adds balancing code and tests for reactions involving compounds as discussed in #651.

@isaacsas @TorkelE

isaacsas commented:
@LalitChauhan56 once we get the other merged and this updated to the merged master I'll give this a review. See my comments on the other PR. Thanks for splitting these up!

LalitChauhan56 commented:
@isaacsas @TorkelE This PR is now up-to-date with master.

TorkelE commented:
Having this in a file called "compound" seems like a poor choice of name. Made rename it "chemistry_functionality"?

LalitChauhan56 commented:
@TorkelE I have changed the file name. Also, should I make a new file for the balancing code part (both in the source code and test files)?

TorkelE commented:
No, that file can contain all the chemistry-related stuff.

LalitChauhan56 commented:
Okay. Also, is the current approach for balancing okay or should I change anything?

TorkelE commented:
Just checking, how many components on the left/right should the balancing be able to take? Is it limited to two species on each side?

E.g. should you be able to balance
```
H + O --> H2O
```
and get 2,1,1? Would it be possible to do
```
H2 + O2 + N2 + C --> H2O + NO2?
```

I will write a couple of additional tests for you to use, next step would be to ensure those are passed.

TorkelE commented:
Some comments:
- `create_matrix` and `get_stoich` have very generic names, but presumably they are meant to be used for balancing equation? Maybe rename them to something more narrow, suggesting what they are doing?
- Could you add some comments for the `create_matrix` function to describe what it is doing. Maybe one before the function, and one before each of the two `for` loops?

Once you confirm whenever you can have any number of species types on each type I will write more tests for you.

LalitChauhan56 commented:
Yes, the conditions you have said above work. Also, there is no limit on the number of reactants and products in the reaction.

LalitChauhan56 commented:
Yes, `create_matrix` and `get_stoich` are used for balancing the equation in the final function `balance` which gives a balanced equation as the output. Okay I will add comments at the specified places.

TorkelE commented:
I'm pasting another 9 reactions here for running tests on. I haven't written out the tests, but I would in each case check that the balanced output is correct, and test that. In each case I have also written what I think would be the correct balanced reaction (but please double check if the algorithm doesn't;t seem to work!).

Several of these cases errors or gives some very strange output, so I would go through them one by one to check what is going on. Please update with any questions once you get them!

```julia
using Catalyst

@variables t
@species C(t) Ca(t) Fe(t) H(t) N(t) Na(t) O(t) P(t) S(t)

# @reaction k, 2H2O --> 2H2O
let 
    @compound H2O(t) 2H O
    r = Reaction(1.0, [H2O], [H2O], [2], [2])
    balance(r)
end

# @reaction k, 23H + 1O --> 7H2O    # Note: Definitely not balanced as given on this line!
let 
    @compound H2O(t) 2H O
    r = Reaction(1.0, [H, O], [H2O], [23, 1], [7])
    balance(r)
end

# @reaction k, 2CH4 + 3O2 --> 2CO2 + 4H2O
let 
    @compound CH4(t) 4H C
    @compound O2(t) 2O
    @compound CO2(t) C 2O
    @compound H2O(t) 2H O
    r = Reaction(1.0, [CH4, O2], [CO2, H2O])
    balance(r)
end

# @reaction k, N2 + 3H2 --> 2NH3
let
    @compound N2(t) 2N
    @compound H2(t) 2H
    @compound NH3(t) N 3H
    r = Reaction(1.0, [N2, H2], [NH3])
    balance(r)
end

# @reaction k, 2C2H5OH + CH3COOH --> 2C4H8O2 + 2H2O
let
    @compound C2H5OH(t) 2C 6H O
    @compound CH3COOH(t) 4C 4H 2O
    @compound C4H8O2(t) 4C 8H 2O
    @compound H2O(t) 2H O
    r = Reaction(1.0, [C2H5OH, CH3COOH], [C4H8O2, H2O])
    balance(r)
end

# @reaction k, 2Ca3PO42 --> 6CaO + 2P4O10
let    
    @compound Ca3PO42(t) 3Ca 2P 8O
    @compound CaO(t) Ca O
    @compound P4O10(t) 4P 10O
    r = Reaction(1.0, [Ca3PO42], [CaO, P4O10])
    balance(r)
end

# @reaction k, 4Fe + 3O2 + 6H2O --> 4FeOH3
let
    @compound O2(t) 2O 
    @compound H20(t) 2H O
    @compound FeOH3(t) Fe 3H O
    r = Reaction(1.0, [Fe, O2, H2O], [FeOH3])
    balance(r)
end

# @reaction k, 2NaOH + H2SO4 --> Na2SO4 + 2H2O
let
    @compound NaOH(t) Na O H
    @compound SO4(t) S 4O
    @compound H2SO4(t) 2H SO4
    @compound Na2SO4(t) 2Na SO4
    @compound H2O(t) 2H O
    r = Reaction(1.0, [NaOH,H2SO4], [Na2SO4,H2O], [], [])
    balance(r)
end

# @reaction k, 4NO2 --> 2N2O4
let
    @compound NO2(t) N 2O
    @compound N2O4(t) 2N 4O
    r = Reaction(1.0, [NO2], [N2O4])
    balance(r)
end
```

TorkelE commented:
Also, maybe the `balance` function should be given a more specific name as well? Maybe `balance_reaction`?

Now if people import Catalyst, if they have anything named `balance` (which is not entirely out of the question) there will be a conflict, which would be unfortunate.

LalitChauhan56 commented:
I tried running these, some of them give weird errors. The problem lies in the approach used within the `get_stoich` function. I somewhat expected it to give errors in some cases and have other approaches just in case (commented in the end of the file). I will figure out the most optimal approach and update the function to better handle such cases.

Also, the balance function expects a reaction with no stoichiometric arrays already attached to it.

For eg-
```
    r = Reaction(1.0, [H, O], [H2O], [23, 1], [7])   
    balance(r) #Expects reaction to be  r = Reaction(1.0, [H, O], [H2O])

```

Also, I'll rename the balance function to be `balance_reaction`

TorkelE commented:
I added the one with the stochiometric array as a edge case, since most reactions will have them. It seems sensible (@isaacsas) to have it work on these as well, but simply ignore the stoichiometries (since these are not important for the balanced equation).

Let's start fixing the errors, also, try incorporating these as tests into the test files as well. Next step would be to make these tests and ensure that they pass.

isaacsas commented:
Yes, this should balance a general reaction, irregardless of whether it has stoichiometry associated with it. You could always make the base version take vectors of substrates and products, and then have a dispatch that works for general reactions to pass through those vectors.

LalitChauhan56 commented:
I fixed the errors and added the the above reactions as tests after correcting some of the commented balanced reactions (reactions 3,4,5).

The main issue was that even after dividing two rational number matrices we sometimes ended up with a float vector instead of a rational number vector and it has to do with how the computer handles division.
For eg - we would sometimes end up with `1.999999999999....` instead of `2` 
I modified the logic to handle this and also to handle the edge cases. 

Also, I added comments to the `create_matrix` function and renamed `balance` to be `balance_reaction`. I would love if you could suggest some better names for the `get_stoich` and `create_matrix` functions.

TorkelE commented:
`balance_reaction_get_stoich`
`balance_reaction_create_matrix`

TorkelE commented:
For me, rounding `1.999999999999....` seems useful. Do you have any idea of how you could modify so that you don't have to deal with floats?

Good to see the test passing :)

isaacsas commented:
```julia
balance_reaction_get_stoich
balance_reaction_create_matrix
```
Those are insanely long names... Why do we need to export these functions at all? Can't we just keep them internal and use the current names?

isaacsas commented:
@LalitChauhan56 I'll take a look at this next week or this weekend -- I'm at a conference this week.

LalitChauhan56 commented:
> Do you have any idea of how you could modify so that you don't have to deal with floats?

To avoid floating point numbers I converted the numbers into rational numbers before division hoping that we would end up as a rational number as a result but in some cases we end up with a floating point number instead. I think this is due to how the computer handles division internally and we sometimes end-up with floating point numbers. However I think dividing the rational numbers actually has more accuracy than dividing two integers directly (i.e. rounding 1.999999999 to 2 should not have any effect). So I still have to use `round()` for such cases whereas normal cases without floats proceed without it. Also, in some cases the previous logic sometimes gave 0 as a stoichiometric coefficient for some cases such as `r = Reaction(1.0, [NaOH,H2SO4], [Na2SO4,H2O])` so I have added logic to handle such cases as well.

LalitChauhan56 commented:
> Why do we need to export these functions at all? Can't we just keep them internal and use the current names?

I think users will rarely use these function as a standalone so I think it makes sense to make them internal.

LalitChauhan56 commented:
Also, just a random thought, would it be useful to have a function which takes in a `ReactionSystem` and returns an array of reactions in the `ReactionSystem` after balancing them/creates a new `ReactionSystem` with the balanced versions of the reactions?

TorkelE commented:
> Also, just a random thought, would it be useful to have a function which takes in a ReactionSystem and returns an array of reactions in the ReactionSystem after balancing them/creates a new ReactionSystem with the balanced versions of the reactions? You could make a `balance_reaction_system` function, which would be quick and would not hurt.

Could you make a MWE of the Float64 conversion? Then we could try and figure out what happens.

TorkelE commented:
Ideally we would want something shorter. Since the issue likely does not depend on Catalyst, the MWE should not have Catalyst as a dependency. We could then use that to see what is going wrong. Often, when trying to build it down to a MWE we actually find the issue and won't need to ask for further help.

I asked Chat GPT and got this description of a MWE:
Creating a Minimal Working Example (MWE) in Julia
Facing a bug in your Julia code? Before diving into forums or seeking help, it's essential to distill your problem into a Minimal Working Example (MWE). An MWE is a stripped-down version of your code that still reproduces the issue. Here's how to craft one:

1. Isolate the Issue
Trim away all the fat. Get rid of any code that isn't directly related to the problem. The goal is to reproduce the bug with as few lines of code as possible.

2. Essential Info Only
If your code relies on specific packages or modules, remember to include the necessary using statements. But, keep it to just what's required.

3. Verify Your MWE
Before sharing your MWE with others, give it a run. Ensure it still throws the same error or displays the issue you're trying to resolve.

4. Context Matters
Alongside your MWE, briefly describe the behavior you expected versus what's actually happening. Mentioning the Julia version can also be a game-changer for those trying to assist.

5. Limit External Dependencies
Your MWE should ideally stand alone. If it needs files or external data, consider replacing them with small, synthetic examples or data snippets.

Why Bother with an MWE?
Creating a Minimal Working Example might feel like an added step, especially when you're eager to resolve a bug. However, it's a time-saver in the long run. Not only does it make it easier for others to help you, but the process of simplifying can often lead you to the root cause of the issue.

LalitChauhan56 commented:
I'll look into this by the end of this weekend, busy with a few training sessions this weekend 😅

LalitChauhan56 commented:
This approach uses rational numbers from the start till the end entirely avoiding float numbers. Also this ensures that the matrix is always square and solved using backward substitution method.

Also, `get_stoich` and `create_matrix` are no longer exported.

TorkelE commented:
Looks good. I might separate the tests for the balancing reaction stuff into a separate test file, e.g. "test/miscellaneous_tests/reaction_balancing.jl" and then add a line to the runtests file:

`@time @safetestset "Reaction balancing" begin include("miscellaneous_tests/reaction_balancing.jl") end`

I think Sam will want to have a look at it as well before we merge, but I think you can start thinking about next steps now.

isaacsas commented:
@LalitChauhan56 you need to update this to master as it currently shows already merged code as new. Can you do that, then I'll take a look.

LalitChauhan56 commented:
I think it is because I changed the name of the file.

isaacsas commented:
Keep the old file name, update to master, then use `git mv` to rename the file.

isaacsas commented:
That should hopefully pick up the changes only.

isaacsas commented on src/chemistry_functionality.jl:
```diff
@@ -0,0 +1,211 @@
+            new_row = zeros(Int, 1, size(A, 2))
+            new_row[end] = 1
+            A = vcat(A, new_row)
+
+            B = zeros(Int64, size(A,1))
+            B[end] = 1
+        end
+
+    # Concatenate B to A to form an augmented matrix   
+    AB = [A B]
+
+    # Apply the Bareiss algorithm
+    ModelingToolkit.bareiss!(AB)
+
+    # Extract the transformed A and B
+    A_transformed = AB[:, 1:end-1]
+    B_transformed = AB[:, end]
+
+    # Convert A_transformed to rational numbers
+    A_transformed_rational = Rational.(A_transformed)
+
+    # Convert B_transformed to rational numbers
+    B_transformed_rational = Rational.(B_transformed)
+
+    # Perform backward substitution
+    X = backward_substitution(A_transformed_rational,B_transformed_rational)
+
+    # Get the denominators of the rational numbers in X
+    denominators = denominator.(X)
+
+    # Compute the LCM of the denominators
+    lcm_value = reduce(lcm, denominators)
+
+    # Multiply each element in X by the LCM of the denominators
+    X_multiplied = X .* lcm_value
+
+    # Convert the rational numbers to integers
+    X_integers = numerator.(X_multiplied)
+
+    return abs.(X_integers)
```
@LalitChauhan56 do you have a reference for this approach? When do you only add one row when m != n? What if m = n + k with k > 1?

More generally, why not use the `nullspace` function from ModelingToolkit that we use for conservation laws, see https://github.com/SciML/Catalyst.jl/blob/e91b001dbaa9ccfe855a54ed79661b332d6b05ae/src/networkapi.jl#L1200-L1241 to get an integer nullspace basis?

isaacsas reviewed (commented):
Did you do any examples where `m = n + k` for `k > 1`?

LalitChauhan56 commented:
We end up 3 types of cases when we are building the matrix before solving -
1. We have equal number of equations and number of variables (m == n)
2. We have more number of equations than the number of variables (m > n)
3. We have more number of variables than the number of equations (m < n)

In the first case it is a straightforward process to solve for X. (using the Bareiss algorithm)

In the second case we slice the matrix containing the equations so that the number of equations is equal to the number of variables and solve it similar to the first case.

In the third case, where we have more variables than the number of equations we define an auxiliary equation
by arbitrarily choosing a value for one of the coefficients ( assuming it to be 1) and solve the system similar to the first case.
Also, assuming the value of one coefficient seems to be enough in solving all systems so adding one row seems to do the job. I tried it with multiple [reactions](https://www.albert.io/blog/balancing-chemical-equations-practice-and-review/) and also added those as tests. Tests now contain all the 3 cases mentioned above.
Also, regarding the use of `nullspace`, I don't think we can use it in the third case because the we assume the equation to be in the form x = 1 so the b vector in Ax= b is not 0. (Correct me if I am wrong).

Also, please let me know if I am missing in case where adding one row is not enough (I couldn't find any).
Reference : http://mathgene.usc.es/matlab-profs-quimica/reacciones.pdf (page 1, last paragraph)

isaacsas commented:
@LalitChauhan56 take a look at the second equation on http://logical.ai/chemistry/html/chem-nullspace.html. It is a case where there are two more columns then rows, but there should be a solution to balancing the reaction (in fact there are an infinite number of solutions, but I think we'd just want to pick one).

LalitChauhan56 commented:
Yes, I think using `nullspace` makes sense. I got confused about the Ax = b part because of the previous approach. I have implemented the nullspace approach now. But in the cases where there are infinite solutions how do we select the appropirate set of

LalitChauhan56 commented:
Sorry that was a misclick.

isaacsas commented:
Maybe it would make sense to return a vector of `Reaction`s when the nullspace dimension is greater than 1 (one for each basis element), along with printing a message that there are multiple independent solutions?

isaacsas commented:
Finishing in https://github.com/SciML/Catalyst.jl/pull/667
